### ISSUE TITLE
`from_thread.run(_sync)?` fails in any thread running Trio

### ISSUE THREAD

**Issue by gschaffner** [python-trio/trio#2534]
hi! `from_thread.run(_sync)?` uses a heuristic to prevent the user from accidentally deadlocking their event loop by writing code like https://github.com/python-trio/trio/pull/1574/files#r435039316.

the heuristic is currently much more aggressive than is required to prevent this deadlock, however. it has the side-effect[^1] of preventing one from calling `from_thread.run(_sync)?` in _any_ thread running trio, _even though such calls are almost always safe and deadlock-free_.

[^1]: citing the `RuntimeError`s reported in `from_thread.run(_sync)?`'s docstrings, this would appear to be an unintended side-effect of the current heuristic.

it would be useful to be able to call `from_thread.run(_sync)?` from trio threads[^2], but the current heuristic blindly prevents this. i'd like to propose relaxing `from_thread.run(_sync)?` to allow calling `from_thread.run(_sync)?` from trio threads. (one use-case of this is https://github.com/agronholm/anyio/issues/525.)

[^2]: for example in order to call `Nursery.start_soon` in a thread-safe manner.

**Comment by oremanj** [comment]
I agree that the check is more zealous than it strictly needs to be. I don't think the idea of multiple Trio threads was contemplated when `trio.from_thread` was written, and even if you loosen the check, `from_thread` is still not a good fit for Trio-to-Trio calls (because it will block the originating thread until the target thread finishes doing whatever you asked it to do). I think a different interface that is designed for cross-Trio-thread calls (with the entrypoint functions being async instead of synchronous) would be clearer; what do you think?

I'm curious what you're doing that requires multiple Trio threads anyway -- we've considered the possibility but hadn't come across anyone with a concrete use case!

**Comment by smurfix** [comment]
Isn't it fairly trivial to cross call from one Trio thread to the other?
`await trio.to_thread(functools.partial(trio.from_thread.run, fn, *args, trio_token=other_task.token))`

If you don't want the overhead, well, creating something like `trio._threads.cross_thread.run(fn, *args)` should be fairly simple.

However IMHO relaxing the rules about not calling any `trio.from_thread.*` function from within Trio isn't the right way to do this.

**Comment by oremanj** [comment]
@gschnaffer thank you for the extremely detailed response! I'm convinced :-)

Would you also benefit from `trio.run(..., allow_nested=True)`? That would create an inner Trio event loop that blocks the outer one until it's complete, with no need for a thread. We don't allow that currently, but there's no principled reason for it, besides the result of nested `trio.run()` being confusing (some people would expect the outer event loop to continue running) and no one's wanted it yet. It sounds like it might solve your problem without the need for an extra thread.

I agree that greenback is somewhat dangerous in that it can violate "no task switches can occur in a synchronous function" assumptions. I don't think it is particularly hazardous otherwise, unless you consider greenlets in general to be hazardous; the async support in SQLAlchemy uses the same technique, and I use greenback in production at my day job.
